server/src/main/java/org/elasticsearch/cluster/coordination/JoinTaskExecutor.java

```diff
@@ -143,19 +148,43 @@ public JoinTaskExecutor(AllocationService allocationService, Logger logger, Rero
                     nodesChanged = true;
                     minClusterNodeVersion = Version.min(minClusterNodeVersion, node.getVersion());
                     maxClusterNodeVersion = Version.max(maxClusterNodeVersion, node.getVersion());
+                    // TODO do we need to check node.isMasterNode here for eligibility? I think since in the checks later master
+                    // eligibility will be taken into account anyway, here we may not need the check?
+                    joiniedNodeNameIds.put(node.getName(), node.getId());
                 } catch (IllegalArgumentException | IllegalStateException e) {
                     results.failure(joinTask, e);
                     continue;
                 }
             }
             results.success(joinTask);
         }
+
         if (nodesChanged) {
             rerouteService.reroute("post-join reroute", Priority.HIGH, ActionListener.wrap(
                 r -> logger.trace("post-join reroute completed"),
                 e -> logger.debug("post-join reroute failed", e)));
 
-            return results.build(allocationService.adaptAutoExpandReplicas(newState.nodes(nodesBuilder).build()));
+            if (joiniedNodeNameIds.isEmpty() == false) {
+                Set<CoordinationMetaData.VotingConfigExclusion> currentVotingConfigExclusions = currentState.getVotingConfigExclusions();
+                Set<CoordinationMetaData.VotingConfigExclusion> newVotingConfigExclusions = currentVotingConfigExclusions.stream()
+                    .map(e -> {
+                        // Update nodeId in VotingConfigExclusion when a new node with excluded node name joins
+                        if (CoordinationMetaData.VotingConfigExclusion.MISSING_VALUE_MARKER.equals(e.getNodeId()) &&
+                            joiniedNodeNameIds.containsKey(e.getNodeName())) {
+                            return new CoordinationMetaData.VotingConfigExclusion(joiniedNodeNameIds.get(e.getNodeName()), e.getNodeName());
+                        } else {
+                            return e;
+                        }
+                    }).collect(Collectors.toSet());
+
+                CoordinationMetaData.Builder coordMetaDataBuilder = CoordinationMetaData.builder(currentState.coordinationMetaData())
```

Comment:
I think we should only do this if we actually resolved any missing values (i.e. `newVotingConfigExclusions` is different from `currentVotingConfigExclusions`).

Reply:
Done.